### ISSUE TITLE
Monkeytype creates circular imports

### ISSUE THREAD

**Issue by bluebird75** [Instagram/MonkeyType#111]
With MonkeyType 18.8.0 and Python 3.6 .

I ran into this on a complex example, but it's actually quite easy to reproduce.

```
d:\work\pyann\monkeytype\toto>type a.py
class A:
    def __init__(self):
        self.titi = 'A_titi'

    def copy_from(self, other):
        self.titi = other.titi

d:\work\pyann\monkeytype\toto>type b.py
from a import A

class B(A):
    def __init__(self):
        super().__init__()
        self.titi = 'B_titi'

d:\work\pyann\monkeytype\toto>type main.py
from a import A
from b import B

if __name__ == '__main__':
    a = A()
    print( a.titi )
    b = B()
    a.copy_from( b )
    print( a.titi )

d:\work\pyann\monkeytype\toto>python main.py
A_titi
B_titi

d:\work\pyann\monkeytype\toto>monkeytype run main.py
A_titi
B_titi

d:\work\pyann\monkeytype\toto>monkeytype list-modules
a
b

d:\work\pyann\monkeytype\toto>monkeytype apply a

d:\work\pyann\monkeytype\toto>monkeytype apply b
Traceback (most recent call last):
  File "c:\program files (x86)\python36-32\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\program files (x86)\python36-32\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Program Files (x86)\Python36-32\Scripts\monkeytype.exe\__main__.py", line 9, in <module>
  File "c:\program files (x86)\python36-32\lib\site-packages\monkeytype\cli.py", line 377, in entry_point_main
    sys.exit(main(sys.argv[1:], sys.stdout, sys.stderr))
  File "c:\program files (x86)\python36-32\lib\site-packages\monkeytype\cli.py", line 362, in main
    handler(args, stdout, stderr)
  File "c:\program files (x86)\python36-32\lib\site-packages\monkeytype\cli.py", line 130, in apply_stub_handler
    stub = get_stub(args, stdout, stderr)
  File "c:\program files (x86)\python36-32\lib\site-packages\monkeytype\cli.py", line 105, in get_stub
    traces.append(thunk.to_trace())
  File "c:\program files (x86)\python36-32\lib\site-packages\monkeytype\encoding.py", line 183, in to_trace
    function = get_func_in_module(self.module, self.qualname)
  File "c:\program files (x86)\python36-32\lib\site-packages\monkeytype\util.py", line 38, in get_func_in_module
    func = get_name_in_module(module, qualname)
  File "c:\program files (x86)\python36-32\lib\site-packages\monkeytype\util.py", line 75, in get_name_in_module
    obj = importlib.import_module(module)
  File "c:\program files (x86)\python36-32\lib\importlib\__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 978, in _gcd_import
  File "<frozen importlib._bootstrap>", line 961, in _find_and_load
  File "<frozen importlib._bootstrap>", line 950, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 655, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 205, in _call_with_frames_removed
  File "d:\work\pyann\monkeytype\toto\b.py", line 2, in <module>
    from a import A
  File "d:\work\pyann\monkeytype\toto\a.py", line 2, in <module>
    from b import B
ImportError: cannot import name 'B'
```

Type information can not be applied to b.py because a circular import has been created in a.py :

```
d:\work\pyann\monkeytype\toto>type a.py
from b import B
class A:
    def __init__(self) -> None:
        self.titi = 'A_titi'

    def copy_from(self, other: B) -> None:
        self.titi = other.titi
```

I am not sure that there is a way around this. The structure of the modules is so that they don't have circular imports without type information but adding it creates the cycle.

In this precise case, the cycle is easy to break. Just by recording a call to A.copy_from() with an A instance as parameter, monkeytype figured out that type A is actually expected and did not use B anymore.

In my complex use case at work, the circular dependencies created are much more complex and can not be solved without deep involvement and probably some refactoring.

**Comment by mpage** [comment]
Thanks for the report! I'm not sure what the right thing to do here is. I'm going to leave this open until we figure out the best course of action going forwards.

**Comment by carljm** [comment]
> I need to put the B annotation in a string to make it work.

Good point. This will no longer be necessary with Python 3.7 if you use `from __future__ import annotations` -- that will stop annotations from being evaluated at runtime.

> I was under the impression that usingmonkeytype applywas a 5 minutes job. Actually, it requires thorough code review. 

The README currently says "MonkeyType's annotations are an informative first draft, to be checked and corrected by a developer." We could definitely expand on this in the docs.

> Pointing to the TYPE_CHECKING trick and string-style type annotation is also important.

Definitely; perhaps this belongs in the FAQ?

> This also raises the question of whether it is a good idea to be precise (method was called with instances of B, B1, B2 so let's annotate that) or to be more generic, like figuring out that B, B1 and B2 have a common ancestor and this is certainly the one needed here.

Yes, there are lots of such questions that come up that we can't be fully confident we are making the correct decision for all cases. We have intentionally chosen to keep the default behavior of MonkeyType very simple and avoid throwing away precision, allowing the developer reviewing the annotations to take all the information into account. But this is exactly why we have the type rewriter system: if you prefer for your project to have MonkeyType automatically simplify `Union[B, B1, B2]` to their common ancestor `A`, it is only a few lines of code in your config to implement that as a type rewriter.

**Comment by ChihweiLHBird** [comment]
> It would probably make sense for us to just always add all new imports inside an `if TYPE_CHECKING` block, right? By definition they are typecheck-only imports. That would solve this problem.
> 
> It probably also provides a reasonable workaround for you in the meantime? E.g. in your example case, if you just changed `a.py` to make the import conditional before trying to apply to `b.py`, the problem would go away.

It would be great if you can provide a option to let MonkeyType do it.

**Comment by GameDungeon** [comment]
Could you guys please add that option?

**Comment by carljm** [comment]
I'm not going to be able to add that option anytime soon, but I would be happy to review pull requests for it.

**Comment by carljm** [comment]
Closed by #282
